kubeadm: add separate page for configuring / migrating cgroup driver (#26786)
kubernetes/website: content/en/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver.md

---
title: Configuring a cgroup driver
content_type: task
weight: 10
---

<!-- overview -->

This page explains how to configure the kubelet cgroup driver to match the container
runtime cgroup driver for kubeadm clusters.

## {{% heading "prerequisites" %}}

You should be familiar with the Kubernetes
[container runtime requirements](/docs/setup/production-environment/container-runtimes).

<!-- steps -->

## Configuring the container runtime cgroup driver

The [Container runtimes](/docs/setup/production-environment/container-runtimes) page
provides details on how to setup a number of different container runtimes with the
`systemd` driver by default.

> The same page explains that the `systemd` driver is recommended for kubeadm based setups

Reviewer: What about moving this sentence and the beginning of the paragraph so first we are telling the user they should use systemd and why, and then we explain how.

Author: makes sense.

instead of the `cgroupfs` driver, because kubeadm manages the kubelet as a systemd service.

## Configuring the kubelet cgroup driver

kubeadm allows you to pass a `KubeletConfiguration` structure during `kubeadm init`.
This `KubeletConfiguration` can include the `cgroupDriver` field which controls the cgroup
driver of the kubelet.

{{< note >}}
If the user is not setting the `cgroupDriver` field under `KubeletConfiguration`,
`kubeadm init` will default it to `systemd`.

> {{< /note >}}

Author: this is new to kubeadm 1.21

Reviewer: Let's add "starting for v1.21" and "for older releases, the user should executed following steps manually"

Author: i think the point of the versioned docs (this PR is targeting dev-1.21) is that all details apply to the target release - i.e. 1.21.

maybe we can use some sort of version tag for this.
will defer to sig-docs for this question.

Author: added: `{{< feature-state for_k8s_version="v1.21" state="stable" >}}` inside the note.
not sure if that is the preferred way.

A minimal example of configuring the field explicitly:

```yaml
# kubeadm-config.yaml
kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta2
kubernetesVersion: v1.21
---
kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1
cgroupDriver: systemd
```

Such a configuration file can then be passed to the kubeadm command:

```shell
kubeadm init --config kubeadm-config.yaml
```

{{< note >}}
Kubeadm uses the same `KubeletConfiguration` for all nodes in the cluster.

> The `KubeletConfiguration` is stored in a [ConfigMap](docs/concepts/configuration/configmap)

Reviewer 2: ```suggestion
The `KubeletConfiguration` will be stored in a [ConfigMap](docs/concepts/configuration/configmap)
```

Author: this shifts the paragraph from present tense to future tense.

Reviewer 2: I was trying to express the future tense.. since the "store" happens after one of the api server is running.

> object under the `kube-system` namespace.

Reviewer 2: ```suggestion
object under the `kube-system` namespace after the control plane is up.
```

Author: IMO this feels more confusing than clarifying.
to be precise it is stored after the first API server's health reports 200.

Reviewer 2: yeah, your description is more accurate. We can leave it as your original wording.

Executing the sub commands `init`, `join` and `upgrade` would result in kubeadm
writing the `KubeletConfiguration` as a file under `/var/lib/kubelet/config.yaml`
and passing it to the local node kubelet.
{{< /note >}}

## Using the `cgroupfs` driver

As this guide explains using the `cgroupfs` driver with kubeadm is not recommended.

To continue using `cgroupfs` and to prevent `kubeadm upgrade` from modifying the
`KubeletConfiguration` cgroup driver on existing setups, you must be explicit
about its value. This applies to a case where you do not wish future versions
of kubeadm to apply the `systemd` driver by default.

> See the bellow section on "Modify the kubelet ConfigMap" for details on

Reviewer: ```suggestion
See the below section on "Modify the kubelet ConfigMap" for details on
```
nit

how to be explicit about the value.

If you wish to configure a container runtime to use the `cgroupfs` driver,
you must refer to the documentation of the container runtime of your choice.

## Migrating to the `systemd` driver

To change the cgroup driver of an existing kubeadm cluster to `systemd` in-place,
a similar procedure to a kubelet upgrade is required. This must include both

> steps outlined bellow.

Reviewer: ```suggestion
steps outlined below.
```

{{< note >}}
Alternatively, it is possible to replace the old nodes in the cluster with new ones

> that use the `systemd` driver. This requires executing only the first step bellow

Reviewer: ```suggestion
that use the `systemd` driver. This requires executing only the first step below
```

before joining the new nodes and ensuring the workloads can safely move to the new
nodes before deleting the old nodes.
{{< /note >}}

### Modify the kubelet ConfigMap

- Find the kubelet ConfigMap name using `kubectl get cm -n kube-system | grep kubelet-config`.
- Call `kubectl edit cm kubelet-config-x.yy -n kube-system` (replace `x.yy` with
the Kubernetes version).
- Either modify the existing `cgroupDriver` value or add a new field that looks like this:

  ```yaml
  cgroupDriver: systemd
  ```
  This field must be present under the `kubelet:` section of the ConfigMap.

### Update the cgroup driver on all nodes

For each node in the cluster:

> - Drain the node using `kubectl drain <node-name> --ignore-daemonsets`

Reviewer 3: Should we link to https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/ for some details.
Most likely, user may use --force for application pods.

Author: i can include the link.

- Stop the kubelet using `systemctl stop kubelet`

> - Stop the container runtime

Reviewer 3: Sometimes, we need to clean up the container runtime dir.
Not sure if this is needed.

Author: from my tests this was not needed.
we can update the guide after this PR merges if we find gaps.

<!-- steps -->

## Configuring the container runtime cgroup driver

The [Container runtimes](/docs/setup/production-environment/container-runtimes) page
explains that the `systemd` driver is recommended for kubeadm based setups instead
of the `cgroupfs` driver, because kubeadm manages the kubelet as a systemd service.

The page also provides details on how to setup a number of different container runtimes with the
`systemd` driver by default.

## Configuring the kubelet cgroup driver

kubeadm allows you to pass a `KubeletConfiguration` structure during `kubeadm init`.
This `KubeletConfiguration` can include the `cgroupDriver` field which controls the cgroup
driver of the kubelet.

{{< note >}}

{{< feature-state for_k8s_version="v1.21" state="stable" >}}

If the user is not setting the `cgroupDriver` field under `KubeletConfiguration`,
`kubeadm init` will default it to `systemd`.
{{< /note >}}

A minimal example of configuring the field explicitly:

```yaml
# kubeadm-config.yaml
kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta2
kubernetesVersion: v1.21
---
kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1

> cgroupDriver: systemd

Reviewer 4: https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc uses 

```
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
nodeRegistration:
  kubeletExtraArgs:
    cgroup-driver: "cgroupfs"
```

to define the cgroupfs.

Confusing...

Author: don't use kubeletExtraArgs, because the cgroup-driver flag is deprecated.
this new page has the right config example.

Author: there are also the support forums and #kubeadm on the k8s slack for users that need help:
https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

i will send a PR to remove the kubeletExtraArgs examples in https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc

Reviewer 4: > there are also the support forums and #kubeadm on the k8s slack for users that need help:
https://github.com/kubernetes/kubernetes/blob/master/SUPPORT.md

I know. But the docs should be the first place to search, shouldn't it?

> i will send a PR to remove the kubeletExtraArgs examples in https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2?utm_source=godoc

:+1:
